setup.go

```diff
@@ -272,21 +271,41 @@ func setup(c *caddy.Controller) error {
 			return errors.New("insecure schemes are only allowed to localhost upstreams")
 		}
 
-		// TODO: remove homebrewed Dialer when https://go-review.googlesource.com/c/net/+/111135 gets merged
-		proxy.RegisterDialerType("https", func(u *url.URL, _ proxy.Dialer) (proxy.Dialer, error) {
+		registerHTTPDialer := func(u *url.URL, _ proxy.Dialer) (proxy.Dialer, error) {
 			// CONNECT request is proxied as-is, so we don't care about target url, but it could be
 			// useful in future to implement policies of choosing between multiple upstream servers.
 			// Given dialer is not used, since it's the same dialer provided by us.
-			return NewHTTPDialer(dialer, true, upstreamURL), nil
-		})
-		proxy.RegisterDialerType("http", func(u *url.URL, _ proxy.Dialer) (proxy.Dialer, error) {
-			return NewHTTPDialer(dialer, false, upstreamURL), nil
-		})
+			d, err := httpclient.NewHTTPConnectDialer(upstreamURL.String())
+			if err != nil {
+				return nil, err
+			}
+			d.Dialer = *dialer
+			if isLocalhost(upstreamURL.Hostname()) && upstreamURL.Scheme == "https" {
+				log.Println("Localhost upstream detected, disabling verification of TLS certificate")
```

Comment:
I assume this is for testing?  It would be good to have  a comment explaining why this case exists.

Reply:
Done